[TT-10294] docs: Tyk 5.13.0 release notes (#1839)
TykTechnologies/tyk-docs: developer-support/release-notes/gateway.mdx

Configure error overrides at multiple levels - globally, per API, or per specific endpoint - with flexible matching by status codes, error types, message patterns, or response content. Support for dynamic templates with validation context enables RFC 7807 compliance and actionable error feedback, while seamless integration with existing analytics and logging ensures complete observability of overridden responses.
Enhanced Developer Experience & Operational Excellence
This release brings significant improvements to day-to-day operations with configurable rate limit headers that now correctly display rate limit information (rather than quota data) in HTTP 429 responses, addressing industry standard compliance. Enhanced session management provides automatic cleanup of expired Redis objects, while TLS certificate rotation support eliminates the need for Gateway restarts during routine certificate updates.

> Additional quality-of-life improvements include underscore support in Vault/Consul secret paths, configurable API compression limits, and numerous stability fixes addressing CORS preflight handling, analytics generation for OAS mock endpoints, and client mTLS authentication between Gateways.

Reviewer: Why is this here? The point of this section is "highlights" not "summary of the entire change log..." 🙈

For a comprehensive list of changes, please refer to the detailed [changelog](#Changelog-v5.13.0) below.

> #### Breaking Changes

Reviewer: ```suggestion
#### Breaking Changes

**Default maximum TLS version** is now inherited from the underlying Golang library and so will be TLS 1.3 unless overridden using `TYK_GW_HTTPSERVEROPTIONS_MAXVERSION`. Previously this defaulted to TLS 1.2. See [here](/api-management/implement-tls#controlling-tls-version-&-cipher-suites) for details how to control TLS version and cipher suites.
```

Added support for TLS certificate rotation in External Service configurations without requiring Gateway restarts. This enhancement allows organizations with strict certificate rotation policies to update certificates seamlessly during runtime.

Key improvements include:

- Hot-reload capability for TLS certificates in External Service configurations, eliminating the need for Gateway pod restarts during certificate rotation
- New optional configuration toggle allowing the Gateway to intelligently reuse mTLS certificates already configured for upstream connections when making OAuth external service calls
- Reduced operational overhead by removing the requirement for infrastructure changes and custom image rebuilds during routine certificate rotation
- Enhanced certificate management flexibility, supporting scenarios where multiple certificates may be configured for different external service endpoints

This feature is particularly valuable for environments with short certificate validity periods (e.g., 90-day rotation cycles) and maintains full backward compatibility with existing External Service configurations.
</Accordion>

> <Accordion title='Added Underscore Support in Vault and Consul Secret Reference Paths'>

Reviewer: The release note section in the ticket explicitly says that we must acknowledge the OSS contributor in the release notes.

---

## 5.13 Release Notes 

### 5.13.0 Release Notes 

#### Release Date 19 May 2026

#### Release Highlights

> Tyk 5.13.0 introduces MCP (Model Context Protocol) Gateway support, bringing AI agent tool servers into Tyk's API management platform. This release also completes Tyk's observability stack with full OpenTelemetry metrics export, and delivers a comprehensive API error response customisation system.

Reviewer: ```suggestion
Tyk 5.13.0 introduces MCP (Model Context Protocol) Gateway support, bringing AI agent tool servers into Tyk's API management platform. This release also completes Tyk's observability stack with full OpenTelemetry metrics export, and delivers a comprehensive API error response customization system.
```

**Gateway support for MCP**

> We have introduced the MCP (Model Context Protocol) Gateway, enabling teams to proxy, secure, and manage AI agent tool servers within Tyk API management. MCP Gateway capabilities are generally available in this release.

Reviewer: ```suggestion
We have introduced the MCP (Model Context Protocol) Gateway, enabling teams to proxy, secure, and manage AI agent tool servers within Tyk API management.

The MCP Gateway is not a new component - the capabilities are generally available in this release of Tyk Gateway.
```

MCP Proxies support Server-Sent Events (SSE) streaming, enabling long-lived agent sessions that survive beyond the configured write timeout. A configurable idle timeout terminates stale SSE streams cleanly, and the Gateway sends a structured error event to the client when an upstream connection drops unexpectedly.

Tool-Based Access Control (TBAC) enforces fine-grained access rules at two levels: JSON-RPC method level (e.g., `tools/call`, `resources/read`) and MCP primitive level (individual tool, resource, or prompt by name).

> Per-primitive rate limiting sets independent rate limits for each method or named primitive within a key or policy. When multiple policies apply, access lists merge as a union and rate limits use the most permissive value.

Reviewer: ```suggestion
Per-primitive rate limiting sets independent rate limits for each method or named primitive within a Session or Policy. When multiple Policies apply, access lists merge as a union and rate limits use the most permissive value.
```

#### Changelog
<a id="Changelog-v5.13.0" data-scroll-offset></a>

##### Changed

<AccordionGroup>

<Accordion title='Upgrade OpenTelemetry SDK dependencies'>
We have updated all OpenTelemetry Go SDK dependencies to their latest compatible versions, addressing significant version skew across core packages, OTLP exporters, and contrib instrumentation libraries.

The upgrade delivers cumulative performance improvements to span lifecycle methods, trace context propagation, metrics recording, and attribute handling that benefit hot paths exercised on every HTTP request. This change maintains full backward compatibility with existing functionality and serves as a prerequisite for upcoming metrics features.
</Accordion>

</AccordionGroup>

##### Added

<AccordionGroup>

<Accordion title='Add MCP Gateway'>
We have introduced the MCP (Model Context Protocol) Gateway, enabling teams to proxy, secure, and manage AI agent tool servers using Tyk API management. MCP Proxy definitions use OpenAPI Specification (OAS) format and support the JSON-RPC over HTTP transport used by MCP clients and servers.

This release includes:

> - **SSE streaming**: MCP Proxies support long-lived Server-Sent Events connections. The write deadline is cleared for SSE streams so connections survive beyond the configured write timeout. A configurable [sse_write_timeout](/tyk-oss-gateway/configuration#sse_write_timeout) idle timeout terminates stale connections cleanly. A structured `event: error` message is sent to the client if the upstream connection drops unexpectedly.

Reviewer 2: `sse_write_timeout` I cannot find this config in the docs. Also, it is not present in the gateway config [PR](https://github.com/TykTechnologies/tyk-docs/pull/2018/changes)

Initially, an invalid URL was configured `/ai-management/mcp-gateway/configuration#sse_write_timeout`

> - **SSE streaming**: MCP Proxies support long-lived Server-Sent Events connections. The write deadline is cleared for SSE streams so connections survive beyond the configured write timeout. A configurable [sse_write_timeout](/ai-management/mcp-gateway/configuration#sse_write_timeout) idle timeout terminates stale connections cleanly. A structured `event: error` message is sent to the client if the upstream connection drops unexpectedly.

Reviewer 2: ```suggestion
- **SSE streaming**: MCP Proxies support long-lived Server-Sent Events connections. The write deadline is cleared for SSE streams so connections survive beyond the configured write timeout. A configurable [sse_write_timeout](/tyk-oss-gateway/configuration#sse_write_timeout) idle timeout terminates stale connections cleanly. A structured `event: error` message is sent to the client if the upstream connection drops unexpectedly.
```

When multiple policies apply to a key, access control lists merge using union semantics: a primitive allowed by any policy is accessible. Block rules take precedence over allow rules. Denied requests receive a JSON-RPC error response rather than an HTTP error, maintaining protocol compatibility with MCP clients.

For configuration details, see the [MCP proxy policies documentation](/ai-management/mcp-gateway/policies).
</Accordion>

<Accordion title='Add per-primitive rate limiting for MCP Proxies'>
MCP Proxies support independent rate limits for each JSON-RPC method and for individual tools, resources, and prompts. Two new fields are available in key and policy access definitions:

- `json_rpc_methods`: rate limits keyed by JSON-RPC method name (e.g., `tools/call`)
- `mcp_primitives`: rate limits keyed by primitive type and name

Limits are translated into Virtual Endpoint Method (VEM) entries at session load time and enforced by the existing rate limiting middleware with no additional processing overhead. The translation is in-memory only and is never persisted to Redis.

> When multiple policies apply to a key, rate limits merge using most-permissive semantics: the highest rate allowed by any policy applies.

Reviewer: ```suggestion
When multiple Policies apply to a Session, rate limits merge using most-permissive semantics: the highest rate allowed by any Policy applies.
```

> For configuration details, see the [MCP proxy policies documentation](/ai-management/mcp-gateway/policies).

Reviewer: ```suggestion
For configuration details, see the [MCP Proxy policies documentation](/ai-management/mcp-gateway/policies).
```

</Accordion>

<Accordion title='Add MCP-specific fields to Gateway access logs'>
Gateway access logs now include MCP-specific fields for requests handled by MCP Proxies:

- `mcp_method`: the JSON-RPC method name (e.g., `tools/call`)
- `mcp_primitive_type`: the primitive category (`tool`, `resource`, or `prompt`)
- `mcp_primitive_name`: the name of the specific tool, resource, or prompt invoked

A new `api_type` field is also added to all access log entries, distinguishing MCP traffic from `classic`, `oas`, and `graphql` API types.

For details, see the [MCP access logs documentation](/ai-management/mcp-gateway/mcp-access-logs).
</Accordion>

<Accordion title='Add MCP dimensions to OpenTelemetry custom metrics'>

> OpenTelemetry custom metrics now expose four MCP-specific dimensions, enabling observability of MCP proxy traffic through configurable metric instruments:

Reviewer: ```suggestion
OpenTelemetry custom metrics now expose four MCP-specific dimensions, enabling observability of MCP Proxy traffic through configurable metric instruments:
```

When an API exceeded its rate limit and returned an HTTP 429 status code, the `X-RateLimit-Limit` and `X-RateLimit-Remaining` response headers were populated from quota data (e.g. monthly limits) rather than rate-limit data (e.g. per-minute thresholds). This was inconsistent with the RFC conventions for these headers and a common source of confusion.
This release adds a new Gateway configuration option [rate_limit_response_headers](/tyk-oss-gateway/configuration#rate_limit_response_headers). When set to "rate_limits", the `X-RateLimit-*` headers are populated from rate-limit data instead of quota data:
- `X-RateLimit-Limit`: the configured rate limit threshold
- `X-RateLimit-Remaining`: remaining requests in the current rate limit window
- `X-RateLimit-Reset`: time until the rate limit window resets

Backward compatibility is preserved: when the option is unset, the headers continue to reflect quota data as before.
</Accordion>

<Accordion title='Simplified management of Session lifetime'>
We have added a new, simplified, approach to configuring [Session lifetime](/api-management/access-control/sessions-and-keys/session-lifecycle) within Redis.

> Two new fields have been added to the Session object, which can be directly configured when creating Sessions using the [Keys API](/api-reference/keys/create-a-key-1) or [Policy API](/api-reference/policies/create-a-policy):

Reviewer 2: ```suggestion
Two new fields have been added to the Session object, which can be directly configured when creating Sessions using the [Keys API](https://tyk.io/docs/api-reference/keys/create-a-key-1) or [Policy API](https://tyk.io/docs/api-reference/policies/create-a-policy):
```

---

## 5.13 Release Notes 

### 5.13.0 Release Notes 

#### Release Date 08 May 2026

#### Release Highlights

Tyk 5.13.0 delivers comprehensive observability, enhanced error handling, and operational excellence, empowering teams to achieve complete visibility into their API ecosystem while streamlining API operations and troubleshooting.

**Complete Observability with OpenTelemetry Metrics**
---
We have introduced comprehensive real-time metrics export through OpenTelemetry, completing Tyk's observability stack alongside existing distributed tracing and logging.

Teams now have complete operational visibility with automatic RED metrics (Rate, Errors, Duration), Go runtime health monitoring, and configuration state tracking.

The new custom metrics framework enables multi-tenant billing, tier-based SLOs, and business KPI tracking through configurable dimensions sourced from request headers, JWT claims, session data, and API metadata. Built-in cardinality controls and metric-to-trace correlation through exemplars ensure production-ready performance while enabling seamless navigation from metric anomalies to specific traces during incident investigation.

> **MCP Gateway**

Reviewer: Not "MCP Proxy Support"?

---
We have introduced the MCP (Model Context Protocol) Gateway, enabling teams to proxy, secure, and manage AI agent tool servers within Tyk API management. MCP Gateway is generally available in this release.

> MCP APIs now support Server-Sent Events (SSE) streaming, enabling long-lived agent sessions that survive beyond the configured write timeout. A configurable idle timeout terminates stale SSE streams cleanly, and the Gateway sends a structured error event to the client when an upstream connection drops unexpectedly.

Reviewer: "now support"? This is new functionality, so there's no previous scenario where they *didn't* support SSE...

Reviewer: ```suggestion
MCP APIs support Server-Sent Events (SSE) streaming, enabling long-lived agent sessions that survive beyond the configured write timeout. A configurable idle timeout terminates stale SSE streams cleanly, and the Gateway sends a structured error event to the client when an upstream connection drops unexpectedly.
```

Are we calling them MCP APIs or MCP Proxies?

Tool-Based Access Control (TBAC) enforces fine-grained access rules at two levels: JSON-RPC method level (e.g., `tools/call`, `resources/read`) and MCP primitive level (individual tool, resource, or prompt by name).

Per-primitive rate limiting sets independent rate limits for each method or named primitive within a key or policy. When multiple policies apply, access lists merge as a union and rate limits use the most permissive value.

> MCP traffic now appears in Gateway access logs with dedicated fields (`mcp_method`, `mcp_primitive_type`, `mcp_primitive_name`) alongside a new `api_type` field present on all access log entries.

Reviewer: Again, don't need "now".

Reviewer: ```suggestion
MCP traffic appears in Gateway access logs with dedicated fields (`mcp_method`, `mcp_primitive_type`, `mcp_primitive_name`) alongside a new `api_type` field present on all access log entries.
```

OpenTelemetry custom metrics expose the same four MCP dimensions, including `mcp_error_code`, aligned with the OpenTelemetry `mcp.*` semantic conventions.

> **Centralized Error Response Customization**

Reviewer: ```suggestion
**Fully Customizable API Error Responses**
```

---
Transform your API's error experience with the new comprehensive error override system. This feature provides complete control over HTTP error responses from both the Gateway and upstream services, enabling standardized, branded error formats that guide users toward solutions rather than leaving them confused.

Configure error overrides at multiple levels - globally and per API - with flexible matching by status codes, error types, message patterns, or response content. Support for dynamic templates with validation context enables RFC 7807 compliance and actionable error feedback, while seamless integration with existing analytics and logging ensures complete observability of overridden responses.

For a comprehensive list of changes, please refer to the detailed [changelog](#Changelog-v5.13.0) below.

#### Breaking Changes

**Default maximum and minimum TLS versions** are now inherited from the underlying Golang library and so will be TLS 1.3 and TLS 1.2, respectively (previously were both set to TLS 1.2).

> You must set [http_server_options.max_version](/tyk-oss-gateway/configuration#http_server_options-max_version) to `771` (or the equivalent environment variable [TYK_GW_HTTPSERVEROPTIONS_MAXVERSION](/tyk-oss-gateway/configuration#http_server_options-max_version)) if you require an upper limit of TLS 1.2.

Reviewer: ```suggestion
You must set [http_server_options.max_version](/tyk-oss-gateway/configuration#http_server_options-max_version) (or the equivalent environment variable)  to `771` if you require an upper limit of TLS 1.2.
```

We have fixed an inconsistent behavior when using the URL rewrite middleware to loop requests using the Tyk protocol (`tyk://api-id/path`).

Previously, query parameters added to the `rewrite_to` URL were silently dropped, while original request parameters were automatically preserved in the looped request.

This behavior was inconsistent with standard HTTP URL rewrites and prevented proper parameter transformation during internal API routing.

**Impact**
- Original request query parameters are no longer automatically forwarded through internal loops
- Existing URL rewrite configurations may lose query parameters that were previously passed through automatically
- APIs relying on automatic parameter forwarding will receive incomplete requests

**Migration Required**
Update your URL rewrite configurations to **explicitly include** any original query parameters you want to preserve. For example:
- **Before:** `"rewrite_to": "tyk://api-123/endpoint"` (original params auto-forwarded)
- **After:** `"rewrite_to": "tyk://api-123/endpoint?param1=$tyk_context.request_data.param1"`

> 

Reviewer: ```suggestion

**Strict Validation of Characters Allowed in Policy IDs**

To avoid an issue where Policy IDs containing special characters could cause problems when parsing API endpoint requests, we have introduced strict validation of Policy IDs during Policy creation and update.

The allowed characters are:
- alphanumeric characters
- `_`
- `-`
- `.`
- `~`

Strict validation can be disabled, if required for existing Policies with incompatible Policy IDs, using the new Gateway configuration `allow_unsafe_policy_ids`. If using this mode, care must be taken not to use characters that could affect URL parsing.

```

Given the potential time difference between your upgrade and the release of this version, we recommend users verify the ongoing support of third-party dependencies they install, as their status may have changed since the release.

#### Deprecations

There are no deprecations in this release.

#### Upgrade instructions
<a id="upgrade-5.13.0"></a>

If you are upgrading to 5.13.0, please follow the detailed [upgrade instructions](#upgrading-tyk).

If your deployment uses MCP Gateway, additional component ordering applies. In hybrid deployments, MDCB 2.11.0 must be deployed before Gateway and Dashboard. Tyk Operator 1.4.0 must be deployed after Gateway and Dashboard are running. Portal 1.17.2 should be deployed before any MCP Proxy policies are created in Dashboard. See [MCP Gateway upgrade considerations](/developer-support/upgrading#mcp-upgrade-considerations) for the full sequence.

#### Downloads

- [Docker Image to pull](https://hub.docker.com/r/tykio/tyk-gateway/tags?page=&page_size=&ordering=&name=v5.13.0)
  - ```bash
    docker pull tykio/tyk-gateway:v5.13.0
    ```
- Helm charts
  - [tyk-charts](/developer-support/release-notes/helm-chart)

> 

Reviewer: ```suggestion

<Note>
Tyk Helm Charts are configured to install the LTS version of Tyk Gateway. You will need to modify them to install v5.13.0.
</Note>

```

##### Added

<AccordionGroup>

<Accordion title='Add MCP Gateway'>
We have introduced the MCP (Model Context Protocol) Gateway, enabling teams to proxy, secure, and manage AI agent tool servers using Tyk API management. MCP Proxy definitions use OpenAPI Specification (OAS) format and support the JSON-RPC over HTTP transport used by MCP clients and servers.

This release includes:
- **SSE streaming**: MCP APIs now support long-lived Server-Sent Events connections. The write deadline is cleared for SSE streams so connections survive beyond the configured write timeout. A configurable [sse_write_timeout](/ai-management/mcp-gateway/configuration#sse_write_timeout) idle timeout terminates stale connections cleanly. A structured `event: error` message is sent to the client if the upstream connection drops unexpectedly.
- **Protected Resource Metadata (PRM)**: The `/.well-known/oauth-protected-resource` endpoint is now served as a middleware in the Gateway chain, making `$tyk_context.*` substitution variables available for use in PRM resource field values.

For configuration details, see the [MCP Gateway documentation](/ai-management/mcp-gateway/overview).
</Accordion>

> <Accordion title='Add Tool-Based Access Control (TBAC) for MCP Proxy APIs'>

Reviewer: Not "now"... and should this perhaps be part of a single entry for MCP support?

Reviewer: Proxy or API?

MCP Proxy APIs now support granular access control at the JSON-RPC method and MCP primitive levels. Two new middleware components enforce access rules independently for each request:

- **JSON-RPC method access control**: Allows or denies requests by method name (e.g., `tools/call`, `resources/read`, `prompts/get`). Configured via `json_rpc_methods_access_rights` in a key or policy access definition.
- **MCP primitive access control**: Allows or denies requests by primitive type (tool, resource, or prompt) and name. Configured via `mcp_access_rights` in a key or policy access definition.

When multiple policies apply to a key, access control lists merge using union semantics: a primitive allowed by any policy is accessible. Block rules take precedence over allow rules. Denied requests receive a JSON-RPC error response rather than an HTTP error, maintaining protocol compatibility with MCP clients.

For configuration details, see the [MCP proxy policies documentation](/ai-management/mcp-gateway/policies).
</Accordion>

<Accordion title='Add per-primitive rate limiting for MCP Proxy APIs'>

> MCP Proxy APIs now support independent rate limits for each JSON-RPC method and for individual tools, resources, and prompts. Two new fields are available in key and policy access definitions:

Reviewer: Again - no previous version, so don't say "now". This should perhaps be part of a single entry for MCP?

- `mcp_primitive_name`: the name of the specific tool, resource, or prompt invoked

A new `api_type` field is also added to all access log entries, distinguishing MCP traffic from `classic`, `oas`, and `graphql` API types.

For details, see the [MCP access logs documentation](/ai-management/mcp-gateway/mcp-access-logs).
</Accordion>

<Accordion title='Add MCP dimensions to OpenTelemetry custom metrics'>
OpenTelemetry custom metrics now expose four MCP-specific dimensions, enabling observability of MCP proxy traffic through configurable metric instruments:

- `mcp_method`: the JSON-RPC method name
- `mcp_primitive_type`: the primitive category (tool, resource, or prompt)
- `mcp_primitive_name`: the name of the specific primitive invoked
- `mcp_error_code`: the JSON-RPC error code when an error is returned; empty otherwise

The dimensions are aligned with the OpenTelemetry `mcp.*` semantic conventions. MCP context is propagated before access control runs, so rejected requests still carry full MCP dimension data.

For details, see the [MCP metrics documentation](/ai-management/mcp-gateway/mcp-metrics).
</Accordion>

> <Accordion title='Add Rate Limit Header Configuration Option'>

Reviewer: ```suggestion
<Accordion title='X-ratelimit Response Header Options'>
```

> We have added a new Gateway configuration option [rate_limit_response_headers](/tyk-oss-gateway/configuration#rate_limit_response_headers) that allows `X-RateLimit-*` headers to be populated from rate limit data instead of quota data. This addresses industry standard compliance and reduces customer confusion when receiving HTTP 429 responses.

Reviewer: We should open by setting the scene - what is currently returned.

Then the fact we've added the new config option and its impact.

Previously, when APIs exceeded their rate limits and returned HTTP 429 status codes, the `X-RateLimit-Limit` and `X-RateLimit-Remaining` headers incorrectly displayed quota information (e.g., monthly limits) rather than rate limit information (e.g., per-minute limits). This violated RFC standards and confused customers who would see remaining quota capacity while receiving rate limit errors.

When [rate_limit_response_headers](/tyk-oss-gateway/configuration#rate_limit_response_headers) is set to `"rate_limits"`, the headers now correctly show:
- `X-RateLimit-Limit`: The configured rate limit threshold
- `X-RateLimit-Remaining`: Remaining requests in the current rate limit window
- `X-RateLimit-Reset`: Time until the rate limit window resets

This change maintains backward compatibility, with the default behavior unchanged when the configuration option is not specified.
</Accordion>

> <Accordion title='Add Configurable Session Object TTL Management for Redis'>

Reviewer: ```suggestion
<Accordion title='Simplified Management of Session Lifetime'>
```

> Tyk now provides improved management of session objects and keys stored in Redis through enhanced TTL (Time To Live) configuration options. This addresses the previous limitation where expired session objects could accumulate indefinitely in Redis storage, requiring manual cleanup scripts.

Reviewer: ```suggestion
We have added a new, simplified, approach to configuring [Session lifetime](/api-management/access-control/sessions-and-keys/session-lifecycle) within Redis.

Two new fields have been added to the Session object, which can be directly configured when creating Sessions using the [Keys API](/api-reference/keys/create-a-key-1) or [Policy API](/api-reference/policies/create-a-policy):
```

> 

Reviewer: ```suggestion
- [post_expiry_action](/tyk-oss-gateway/configuration#post_expiry_action) - determines what happens to the data in Redis after the `expires` timestamp is reached.
- [post_expiry_grace_period](/tyk-oss-gateway/configuration#post_expiry_grace_period) - defines how long (in seconds) the Session is kept in Redis after expiration (if the `post_expiry_actions` is to retain the Session)
```

> Key improvements include:

Reviewer: ```suggestion
```

These bullets are nonsense - they talk about the legacy behaviour/controls.

> - Enhanced [session_lifetime](/tyk-oss-gateway/configuration#session_lifetime) configuration at the API level and [global_session_lifetime](/tyk-oss-gateway/configuration#global_session_lifetime) at the Gateway level to properly manage Redis object TTL

Reviewer: ```suggestion
```

> - Added [session_lifetime_respects_key_expiration](/tyk-oss-gateway/configuration#session_lifetime_respects_key_expiration) setting to ensure session objects are not deleted from Redis before their associated keys have expired

Reviewer: ```suggestion
```

> - Improved coordination between key expiry periods and Redis TTL to prevent premature deletion of valid tokens

Reviewer: ```suggestion
```

> - Automatic cleanup of expired session objects, reducing Redis storage requirements and eliminating the need for manual maintenance scripts

Reviewer: ```suggestion
```

> This provides an enterprise-grade solution for Redis storage management while maintaining backward compatibility with existing key expiration behavior. Session objects will now be automatically removed from Redis when appropriate, based on the configured TTL settings.

Reviewer: ```suggestion
The existing Gateway-wide [global session lifetime](/api-management/access-control/sessions-and-keys/session-lifecycle#gateway-level-settings) override is still respected.

The [legacy API level controls](/api-management/access-control/sessions-and-keys/session-lifecycle#legacy-controls) can still be used if both new fields are set to `0` (or unset) so **there is no change in behavior for existing Sessions**.
```

</Accordion>

> <Accordion title='Add TLS Certificate Rotation Support for External Service Config'>

Reviewer: This is ONLY for the certificates used in the `oauth` service (for integration with external Identity Providers (for client and upstream auth) - not for other external services.

It is ONLY for these fields in the Gateway config:
- `external_services.oauth.mtls.cert_file`
- `external_services.oauth.mtls.key_file`
- `external_services.oauth.mtls.ca_file`

See https://github.com/TykTechnologies/tyk-docs/pull/2064

We have added support for TLS certificate rotation in External Service configurations without requiring Gateway restarts. This enhancement allows organizations with strict certificate rotation policies to update certificates seamlessly during runtime.

Key improvements include:

- Hot-reload capability for TLS certificates in External Service configurations, eliminating the need for Gateway pod restarts during certificate rotation
- New optional configuration toggle allowing the Gateway to intelligently reuse mTLS certificates already configured for upstream connections when making OAuth external service calls
- Reduced operational overhead by removing the requirement for infrastructure changes and custom image rebuilds during routine certificate rotation
- Enhanced certificate management flexibility, supporting scenarios where multiple certificates may be configured for different external service endpoints

This feature is particularly valuable for environments with short certificate validity periods (e.g., 90-day rotation cycles) and maintains full backward compatibility with existing External Service configurations.
</Accordion>

> <Accordion title='Add Underscore Support in Vault and Consul Secret Reference Paths'>

Reviewer: ```suggestion
<Accordion title='Allow Underscores in Vault and Consul Secret Reference Paths'>
```

We have added support for underscore characters in `$secret_vault` and `$secret_consul` secret reference paths. Previously, the underscore character was not supported in these Key-Value store reference paths due to regex pattern limitations, causing references to be truncated at the first underscore. For example, `$secret_vault.kv-v2/path.API_KEY` would only match up to `.API`, silently dropping `_KEY`.

This enhancement aligns `$secret_vault` and `$secret_consul` patterns with existing `$secret_env`, `$tyk_context`, and `$tyk_meta` patterns that already support underscores. Users can now reference secrets using standard naming conventions like `API_KEY` and `DB_PASSWORD` without needing to duplicate secrets under underscore-free names. The change maintains full backward compatibility with existing references that don't contain underscores.
</Accordion>

> <Accordion title='Add Configurable API Compression Size Limit'>

Reviewer: Improve this to be more generic about limiting the size of API defs and policies in the Data Plane cache for distributed deployments (cover this across the title and content).

We seem to have missed TT-16761.

> We have added a new Gateway configuration option [storage.max_decompressed_size](/tyk-oss-gateway/configuration#storage-max_decompressed_size) to control the maximum size limit for API definition decompression. This enhances the API compression feature by allowing operators to adjust the decompression limit based on their deployment requirements.

Reviewer: This must refer to the fact we introduced API def compression in version X and it was previously limited to 100MB.

I would merge this with the fact we can now compress Policies - and the limit is separately applied to each asset type.

> When [storage.max_decompressed_size](/tyk-oss-gateway/configuration#storage-max_decompressed_size) is not configured, the Gateway maintains the default 100MB limit for backward compatibility. When set to a positive value ≥ 1MB, the Gateway uses that value as the decompression cap. Values below 1MB are automatically clamped to 1MB with a warning logged to prevent potential security issues while avoiding service disruption.

Reviewer: This could be said in a few words... the default is 100MB, minimum is 1MB.

> This configuration change applies only to Gateway-level storage settings and does not affect the public API definition schema or Dashboard contracts.

Reviewer: ????

</Accordion>

> <Accordion title='Add Centralized Error Response Customization'>

Reviewer: ```suggestion
<Accordion title='Fully Customizable API Error Responses'>
```

This is a major feature and should sit immediately below MCP and OTel changes, minor feature changes should go below.

> We have added a comprehensive mechanism to customize HTTP error responses generated by both the Tyk Gateway and upstream services. This enhancement allows operators to fully control error outputs to standardize formats, mask sensitive information, and comply with industry standards.

Reviewer: ```suggestion
We have uplifted the existing error response templating system with a new comprehensive mechanism to customize HTTP error responses generated by both the Tyk Gateway and upstream services.

This allows users to fully control error responses to standardize formats, mask sensitive information, and comply with industry standards.
```

**Key Capabilities:**

> - **Multi-level Configuration:** Configure error overrides globally (gateway-wide), or per API with clear precedence rules

Reviewer: ```suggestion
- **Multi-level Configuration:** Configure error overrides globally (Gateway-wide), or per API with clear precedence rules
```

**Template Support:** Response templates support dynamic variables including `{{.StatusCode}}`, `{{.Message}}`, and validation-specific context like `{{.InvalidParams}}` for detailed error feedback.

**Precedence Order:**
1. API-level overrides
2. Gateway-level overrides
3. Default Tyk responses (fallback)

**Supported Error Types:**
- Gateway errors: Authentication failures, rate limits, validation errors, transform failures
- Proxy errors: DNS failures, timeouts, circuit breaker events
- Upstream errors: 4xx/5xx responses from backend services

**Benefits:**
- Standardize error responses across APIs and environments
- Mask sensitive internal error details for security
- Provide consistent error formats for client applications
- Enable RFC 7807 Problem Details compliance
- Reduce client-side error handling complexity

This feature integrates seamlessly with existing analytics, logging, and observability systems; all overridden responses are properly captured in access logs, analytics records, and distributed traces. The system is fully backward compatible with no configuration required; existing error responses remain unchanged unless explicitly overridden.
</Accordion>

> <Accordion title='Add Gateway Go Runtime Metrics via OpenTelemetry'>

Reviewer: Can this be included in the OTel Metrics entry? Why is it separated?

- **Metric-to-Trace Correlation:** Automatic exemplar support links metric anomalies directly to distributed traces, enabling seamless navigation from metrics to traces during incident investigation.
- **Gateway Identity Attributes:** Configurable resource attributes help identify and filter individual Gateway instances in multi-node deployments.

**Configuration:** Metrics are configured under the new [opentelemetry.metrics](/tyk-oss-gateway/configuration#opentelemetry-metrics) object with independent control from tracing

**Default Metrics Exported:**
- **Request Metrics:** `http.server.request.duration`, `tyk.gateway.request.duration`, `tyk.upstream.request.duration`, `tyk.api.requests.total`
- **Runtime Metrics:** Go memory usage, goroutine count, GC performance
- **Configuration Metrics:** Loaded APIs/policies count, reload operations and duration

**Production Controls:** To ensure safe metric collection on the request hot path, the system includes built-in cardinality limits (default: 2,000 combinations per metric) and conditional data loading to prevent performance degradation.
</Accordion>

</AccordionGroup>

> ##### Fixed

Reviewer: Please update with any changes agreed in 5.8.14 release notes.

</Accordion>

<Accordion title='Fix OpenTelemetry Configuration File Settings Being Ignored'>
We have resolved an issue where OpenTelemetry settings in the Gateway configuration file were not being applied. Previously, when users configured OpenTelemetry via the `opentelemetry` section in the Gateway config file (including `enabled`, `exporter`, and `endpoint` fields), these settings were ignored, and only environment variables like `TYK_GW_OPENTELEMETRY_ENABLED` would take effect.

Tyk Gateway now properly reads and applies OpenTelemetry configuration from the config file as documented.
</Accordion>

<Accordion title='Resolve issue with Gateway entering an unresponsive state during registration failures'>
We have fixed an issue where the Gateway would fail to load APIs and policies if the Control Plane database was temporarily unavailable during startup (either directly or via MDCB). The Gateway will now automatically retry loading configurations with exponential backoff until successful, restoring self-healing capabilities without requiring a manual restart.
</Accordion>

<Accordion title='Add Strict Validation for Policy ID Characters'>

> We have resolved an issue where policy IDs containing special characters could cause problems when parsing API endpoint requests. Previously, policy IDs with characters such as `#`, `?`, `%`, and `/` would interfere with URL parsing in Dashboard and Gateway API endpoints that use the policy ID as a path parameter, potentially causing request failures or unexpected behavior.

Reviewer: ```suggestion
We have resolved an issue where Policy IDs containing special characters could cause problems when parsing API endpoint requests. Previously, Policy IDs with characters such as `#`, `?`, `%`, and `/` would interfere with URL parsing in Tyk Gateway API endpoints that use the Policy ID as a path parameter, potentially causing request failures or unexpected behavior.
```

> Strict validation has been introduced across both the Dashboard and Gateway APIs to restrict policy identifiers to a safe character set (alphanumeric characters plus `_`, `-`, `.`, `~`). The validation occurs during policy creation and updates via the following endpoints:

Reviewer: ```suggestion
Strict validation has been introduced to restrict policy identifiers to a safe character set (alphanumeric characters plus `_`, `-`, `.`, `~`). The validation occurs during Policy creation and updates via the following endpoints:
```

> - `POST /api/portal/policies` and `PUT /api/portal/policies/{id}` (Dashboard API)

Reviewer: ```suggestion
```

> - `POST /tyk/policies` and `PUT /tyk/policies/{polID}` (Gateway API)

Reviewer: ```suggestion
- `POST /tyk/policies`
- `PUT /tyk/policies/{polID}`
```

> The API now returns a clear error message if an unacceptable character is used in the policy ID field, ensuring reliable policy management across all deployment types.

Reviewer: ```suggestion
A new Gateway configuration option `allow_unsafe_policy_ids` has been added to disable this character validation in case of existing non-compliant Policy IDs. If you must use other characters in your Policy IDs, be careful to avoid any that might cause URL parsing issues.
```

</Accordion>

<Accordion title='Fix error response format for SOAP+XML requests'>
We have resolved an issue where requests with the `application/soap+xml` Content-Type received JSON-formatted error responses instead of the expected XML format. The Gateway now correctly returns XML-formatted errors for SOAP requests.
</Accordion>

> <Accordion title='Fix Missing Default Configuration for Span Batch Processing'>

Reviewer: Is this a bug from a previous version, or just something that came from new functionality in 5.13.0?

Reviewer 3: This was added in 5.12 "Prevent OpenTelemetry Span Loss in High-Throughput Environments"

We have resolved an issue where SpanBatchConfig was not initialized with proper default values. Previously, this could cause span batch processing to behave unpredictably or fail when no explicit configuration was provided. The Gateway now properly sets appropriate defaults for span batch processing configuration, ensuring consistent behavior across all installations.
</Accordion>

> <Accordion title='Fix Missing API Metrics and Access Logs for Cache Middleware'>

Reviewer: Is this a bug from a previous version, or just something that came from new functionality in 5.13.0?

Reviewer 3: It is true that OTel part is not relevant as a bug fixed, as this is a new feature added in this release. I will rework the entry

We have resolved an issue where cache hits served by the Redis cache middleware were missing observability signals.

Previously, when a response was served from cache, the middleware would short-circuit the request chain and only record analytics, failing to produce access log entries or OpenTelemetry API metrics. This made cached responses invisible to operators monitoring the Gateway.

The cache middleware now properly generates all observability signals for cached responses, ensuring complete visibility into both cache hits and cache misses.
</Accordion>

> <Accordion title='Fix Missing API and Policy Count Metrics When Gateway Fails MDCB Connection'>

Reviewer: Is this a bug from a previous version, or just something that came from new functionality in 5.13.0?

Reviewer 3: Removing this as a standalone Fixed entry

We have resolved an issue where the `tyk.gateway.apis.loaded` and `tyk.gateway.policies.loaded` gauge metrics were not emitted when the Gateway failed to connect to MDCB during startup.

Previously, if the Gateway could not establish an MDCB connection, these metrics would not be reported at all, making it difficult to detect configuration or connection failures. The Gateway now properly emits these metrics with a value of 0 when no APIs or policies are loaded due to MDCB connection failures, providing a clear indicator of startup issues.
</Accordion>

> <Accordion title='Fix VEM path preservation for undefined MCP primitives'>

Reviewer: This looks like something that came from new functionality in 5.13.0 and hence should not be in the change log

Tool-Based Access Control (TBAC) enforces fine-grained access rules at two levels: JSON-RPC method level (e.g., `tools/call`, `resources/read`) and MCP primitive level (individual tool, resource, or prompt by name).

Per-primitive rate limiting sets independent rate limits for each method or named primitive within a key or policy. When multiple policies apply, access lists merge as a union and rate limits use the most permissive value.

MCP traffic now appears in Gateway access logs with dedicated fields (`mcp_method`, `mcp_primitive_type`, `mcp_primitive_name`) alongside a new `api_type` field present on all access log entries.

OpenTelemetry custom metrics expose the same four MCP dimensions, including `mcp_error_code`, aligned with the OpenTelemetry `mcp.*` semantic conventions.

**Centralized Error Response Customization**
---
Transform your API's error experience with the new comprehensive error override system. This feature provides complete control over HTTP error responses from both the Gateway and upstream services, enabling standardized, branded error formats that guide users toward solutions rather than leaving them confused.

> Configure error overrides at multiple levels - globally, per API, or per specific endpoint - with flexible matching by status codes, error types, message patterns, or response content. Support for dynamic templates with validation context enables RFC 7807 compliance and actionable error feedback, while seamless integration with existing analytics and logging ensures complete observability of overridden responses.

Reviewer 4: actually per specific endpoint is not supported yet in this version only globally and per API

</Note>

##### 3rd Party Dependencies & Tools

| Third Party Dependency | Tested Versions | Compatible Versions | Comments | 
| ---------------------- | --------------- | ------------------- | -------- | 
| [Golang](https://go.dev/dl/)               | 1.25                   |  1.25  | [Go plugins](/api-management/plugins/golang) must be built using Go 1.25 | 
| [Redis](https://redis.io/download/)    | 6.2.x, 7.x, 7.4.x      | 6.2.x, 7.x, 7.4.x      | | 
| [Valkey](https://valkey.io/download/)  | 7.2.x, 8.0.x, 8.1.x    | 7.2.x, 8.0.x, 8.1.x    | | 
| [OpenAPI Specification](https://spec.openapis.org/oas/v3.1.2.html) | 3.1.x, 3.0.x | 3.1.x, 3.0.x | Supported by [Tyk OAS](/api-management/gateway-config-tyk-oas)|

Given the potential time difference between your upgrade and the release of this version, we recommend users verify the ongoing support of third-party dependencies they install, as their status may have changed since the release.

#### Deprecations

There are no deprecations in this release.

#### Upgrade instructions
<a id="upgrade-5.13.0"></a>

If you are upgrading to 5.13.0, please follow the detailed [upgrade instructions](#upgrading-tyk).

> If your deployment uses MCP Gateway, additional component ordering applies. In hybrid deployments, MDCB 2.11.0 must be deployed before Gateway and Dashboard. Tyk Operator 1.4.0 must be deployed after Gateway and Dashboard are running. Portal 1.17.1 should be deployed before any MCP Proxy policies are created in Dashboard. See [MCP Gateway upgrade considerations](/developer-support/upgrading#mcp-upgrade-considerations) for the full sequence.

Reviewer 2: ```suggestion
If your deployment uses MCP Gateway, additional component ordering applies:
- In hybrid deployments, MDCB 2.11.0 must be deployed before Gateway and Dashboard. 
- Tyk Operator 1.4.0 must be deployed after Gateway and Dashboard are running. 
- Portal 1.17.1 should be deployed before any MCP Proxy policies are created in Dashboard. 

See [MCP Gateway upgrade considerations](/developer-support/upgrading#mcp-upgrade-considerations) for the full sequence.
```

When [storage.max_decompressed_size](/tyk-oss-gateway/configuration#storage-max_decompressed_size) is not configured, the Gateway maintains the default 100MB limit for backward compatibility. When set to a positive value ≥ 1MB, the Gateway uses that value as the decompression cap. Values below 1MB are automatically clamped to 1MB with a warning logged to prevent potential security issues while avoiding service disruption.

This configuration change applies only to Gateway-level storage settings and does not affect the public API definition schema or Dashboard contracts.
</Accordion>

<Accordion title='Add Centralized Error Response Customization'>
We have added a comprehensive mechanism to customize HTTP error responses generated by both the Tyk Gateway and upstream services. This enhancement allows operators to fully control error outputs to standardize formats, mask sensitive information, and comply with industry standards.

**Key Capabilities:**

> - **Multi-level Configuration:** Configure error overrides globally (gateway-wide), per API, or per specific endpoint with clear precedence rules

Reviewer 4: same here no per specific endpoint support yet

- **Flexible Matching:** Target errors by exact status codes (401, 500), patterns (4xx, 5xx), error flags, message patterns, or response body content
- **Rich Response Customization:** Override status codes, response bodies, headers, and use template files with dynamic variables
- **Universal Coverage:** Customize both Gateway-generated errors (authentication, rate limiting, validation) and upstream service errors (4xx/5xx responses)

**Configuration:** Error overrides use a map-based structure indexed by HTTP status code in the new [error_overrides](/tyk-oss-gateway/configuration#error_overrides) configuration.

**Template Support:** Response templates support dynamic variables including `{{.StatusCode}}`, `{{.Message}}`, and validation-specific context like `{{.invalid_params}}` for detailed error feedback.

> 

Reviewer 4: This is {{.InvalidParams}} (camel case like status code or message vars) not {{.invalid_params}}

**Precedence Order:**

> 1. Endpoint-level overrides (highest priority)

Reviewer 4: same here not endpoint support yet

---

## 5.13 Release Notes 

### 5.13.0 Release Notes 

#### Release Date 08 May 2026

#### Release Highlights

> Tyk 5.13.0 delivers comprehensive observability, enhanced error handling, and operational excellence, empowering teams to achieve complete visibility into their API ecosystem while streamlining API operations and troubleshooting.

Reviewer: This opening paragraph should mention the key highlights - MCP Gateway, Observability, Error Response Customisation - and then be followed by the in depth highlights for each.

I would order those highlights in that order:
MCP
Observability
Error customisation

**Strict Validation of Characters Allowed in Policy IDs**

To avoid an issue where Policy IDs containing special characters could cause problems when parsing API endpoint requests, we have introduced strict validation of Policy IDs during Policy creation and update.

The allowed characters are:
- alphanumeric characters
- `_`
- `-`
- `.`
- `~`

Strict validation can be disabled, if required for existing Policies with incompatible Policy IDs, using the new Gateway configuration `allow_unsafe_policy_ids`. If using this mode, care must be taken not to use characters that could affect URL parsing.

#### Dependencies 
<a id="dependencies-5.13.0"></a>

##### Compatibility Matrix For Tyk Components

| Gateway Version | Recommended Releases | Backwards Compatibility |
|--------|-------------------|---- |
| 5.13.0 | MDCB v2.11.0       | MDCB v2.11.0 |
|        | Operator v1.4.0   | Operator v0.17 |
|        | Sync v2.1.7       | Sync v2.1.0 |
|        | Helm Chart v5.2.0   | Helm all versions |
|        | Pump v1.15.0      | Pump all versions |

<Note>
To use MCP Gateway features introduced in 5.13.0, upgrade the following Tyk components at the same time:

> - **Dashboard v5.13.0**: MCP Proxy management and access control configuration

Reviewer: ```suggestion
- **Dashboard 5.13.0**: MCP Proxy management and access control configuration
```

> - **Pump v1.15.0**: MCP analytics (MongoDB, PostgreSQL, Elasticsearch, and Prometheus backends)

Reviewer: ```suggestion
- **Pump 1.15.0**: MCP analytics (MongoDB, PostgreSQL, Elasticsearch, and Prometheus backends)
```

> - **MDCB v2.11.0**: MCP analytics routing in distributed deployments

Reviewer: ```suggestion
- **MDCB 2.11.0**: MCP analytics routing in distributed deployments
```

> - **Operator v1.4.0**: MCP Proxy management via Kubernetes CRDs

Reviewer: ```suggestion
- **Operator 1.4.0**: MCP Proxy management via Kubernetes CRDs
```

Does Operator's OSS functionality work with MCP Proxies?

> - **Sync v2.1.7**: MCP Proxy support in dump and sync operations

Reviewer: ```suggestion
- **Sync 2.1.7**: MCP Proxy support in `dump` and `sync` operations
```

> - **EDP v1.17.2**: filters all MCP Proxy policies from the developer catalogue; v1.17.1 provides a partial mitigation covering policies with TBAC or per-primitive rate limiting configured

Reviewer: ```suggestion
```

Developer Portal is not relevant to Gateway release notes - it integrates only with Dashboard, not Gateway.

</Note>

##### 3rd Party Dependencies & Tools

| Third Party Dependency | Tested Versions | Compatible Versions | Comments | 
| ---------------------- | --------------- | ------------------- | -------- | 
| [Golang](https://go.dev/dl/)               | 1.25                   |  1.25  | [Go plugins](/api-management/plugins/golang) must be built using Go 1.25 | 
| [Redis](https://redis.io/download/)    | 6.2.x, 7.x, 7.4.x      | 6.2.x, 7.x, 7.4.x      | | 
| [Valkey](https://valkey.io/download/)  | 7.2.x, 8.0.x, 8.1.x    | 7.2.x, 8.0.x, 8.1.x    | | 
| [OpenAPI Specification](https://spec.openapis.org/oas/v3.1.2.html) | 3.1.x, 3.0.x | 3.1.x, 3.0.x | Supported by [Tyk OAS](/api-management/gateway-config-tyk-oas)|

Given the potential time difference between your upgrade and the release of this version, we recommend users verify the ongoing support of third-party dependencies they install, as their status may have changed since the release.

#### Deprecations

There are no deprecations in this release.

#### Upgrade instructions
<a id="upgrade-5.13.0"></a>

If you are upgrading to 5.13.0, please follow the detailed [upgrade instructions](#upgrading-tyk).

> If your deployment uses MCP Gateway, additional component ordering applies. In distributed deployments, MDCB 2.11.0 must be deployed before Gateway and Dashboard. Tyk Operator 1.4.0 must be deployed after Gateway and Dashboard are running. Portal 1.17.2 should be deployed before any MCP Proxy policies are created in Dashboard. See [MCP Gateway upgrade considerations](/developer-support/upgrading#mcp-upgrade-considerations) for the full sequence.

Reviewer: ```suggestion
If your deployment uses MCP Gateway, additional component ordering applies. In distributed deployments, MDCB 2.11.0 must be deployed before Gateway and Dashboard. Tyk Operator 1.4.0 must be deployed after Gateway and Dashboard are running. See [MCP Gateway upgrade considerations](/developer-support/upgrading#mcp-upgrade-considerations) for the full sequence.
```

#### Downloads

- [Docker Image to pull](https://hub.docker.com/r/tykio/tyk-gateway/tags?page=&page_size=&ordering=&name=v5.13.0)
  - ```bash
    docker pull tykio/tyk-gateway:v5.13.0
    ```
- Helm charts
  - [tyk-charts](/developer-support/release-notes/helm-chart)

<Note>
Tyk Helm Charts are configured to install the LTS version of Tyk Gateway. You will need to modify them to install v5.13.0.
</Note>

- [Source code tarball of Tyk Gateway v5.13.0](https://github.com/TykTechnologies/tyk/releases/tag/v5.13.0)

#### Changelog
<a id="Changelog-v5.13.0" data-scroll-offset></a>

##### Changed

<AccordionGroup>

> <Accordion title='Upgrade OpenTelemetry SDK Dependencies'>

Reviewer: ```suggestion
<Accordion title='Upgrade OpenTelemetry SDK dependencies'>
```

We have updated all OpenTelemetry Go SDK dependencies to their latest compatible versions, addressing significant version skew across core packages, OTLP exporters, and contrib instrumentation libraries.

The upgrade delivers cumulative performance improvements to span lifecycle methods, trace context propagation, metrics recording, and attribute handling that benefit hot paths exercised on every HTTP request. This change maintains full backward compatibility with existing functionality and serves as a prerequisite for upcoming metrics features.
</Accordion>

</AccordionGroup>

##### Added

<AccordionGroup>

<Accordion title='Add MCP Gateway'>
We have introduced the MCP (Model Context Protocol) Gateway, enabling teams to proxy, secure, and manage AI agent tool servers using Tyk API management. MCP Proxy definitions use OpenAPI Specification (OAS) format and support the JSON-RPC over HTTP transport used by MCP clients and servers.

This release includes:

> - **SSE streaming**: MCP APIs now support long-lived Server-Sent Events connections. The write deadline is cleared for SSE streams so connections survive beyond the configured write timeout. A configurable [sse_write_timeout](/ai-management/mcp-gateway/configuration#sse_write_timeout) idle timeout terminates stale connections cleanly. A structured `event: error` message is sent to the client if the upstream connection drops unexpectedly.

Reviewer: ```suggestion
- **SSE streaming**: MCP APIs support long-lived Server-Sent Events connections. The write deadline is cleared for SSE streams so connections survive beyond the configured write timeout. A configurable [sse_write_timeout](/ai-management/mcp-gateway/configuration#sse_write_timeout) idle timeout terminates stale connections cleanly. A structured `event: error` message is sent to the client if the upstream connection drops unexpectedly.
```

Again - I don't know why this says "now", as it's new functionality. We're inconsistent with the use of MCP Proxy vs MCP API. It's best to avoid introducing risk of confusion when adding new terminology.

> - **Protected Resource Metadata (PRM)**: The `/.well-known/oauth-protected-resource` endpoint is now served as a middleware in the Gateway chain, making `$tyk_context.*` substitution variables available for use in PRM resource field values.

Reviewer: ```suggestion
- **Protected Resource Metadata (PRM)**: The `/.well-known/oauth-protected-resource` endpoint is served as a middleware in the Gateway chain, making `$tyk_context.*` substitution variables available for use in PRM resource field values.
```

Again - why "now" when referring to new functionality?

For configuration details, see the [MCP Gateway documentation](/ai-management/mcp-gateway/overview).
</Accordion>

<Accordion title='Add Tool-Based Access Control (TBAC) for MCP Proxy APIs'>

> MCP Proxy APIs now support granular access control at the JSON-RPC method and MCP primitive levels. Two new middleware components enforce access rules independently for each request:

Reviewer: ...now... and here we've combined both Proxy and API!

- `mcp_primitive_name`: the name of the specific tool, resource, or prompt invoked

A new `api_type` field is also added to all access log entries, distinguishing MCP traffic from `classic`, `oas`, and `graphql` API types.

For details, see the [MCP access logs documentation](/ai-management/mcp-gateway/mcp-access-logs).
</Accordion>

<Accordion title='Add MCP dimensions to OpenTelemetry custom metrics'>
OpenTelemetry custom metrics now expose four MCP-specific dimensions, enabling observability of MCP proxy traffic through configurable metric instruments:

- `mcp_method`: the JSON-RPC method name
- `mcp_primitive_type`: the primitive category (tool, resource, or prompt)
- `mcp_primitive_name`: the name of the specific primitive invoked
- `mcp_error_code`: the JSON-RPC error code when an error is returned; empty otherwise

The dimensions are aligned with the OpenTelemetry `mcp.*` semantic conventions. MCP context is propagated before access control runs, so rejected requests still carry full MCP dimension data.

For details, see the [MCP metrics documentation](/ai-management/mcp-gateway/mcp-metrics).
</Accordion>

> <Accordion title='Fully Customizable API Error Responses'>

Reviewer: ```suggestion
<Accordion title='Fully customizable API error responses'>
```

We have uplifted the existing error response templating system with a new comprehensive mechanism to customize HTTP error responses generated by both the Tyk Gateway and upstream services.

This allows users to fully control error responses to standardize formats, mask sensitive information, and comply with industry standards.

> **Key Capabilities:**

Reviewer: ```suggestion
**Key capabilities:**
```

</Accordion>

<Accordion title='X-RateLimit response headers can now reflect rate-limit values'>
When an API exceeded its rate limit and returned an HTTP 429 status code, the `X-RateLimit-Limit` and `X-RateLimit-Remaining` response headers were populated from quota data (e.g. monthly limits) rather than rate-limit data (e.g. per-minute thresholds). This was inconsistent with the RFC conventions for these headers and a common source of confusion.
This release adds a new Gateway configuration option [rate_limit_response_headers](/tyk-oss-gateway/configuration#rate_limit_response_headers). When set to "rate_limits", the `X-RateLimit-*` headers are populated from rate-limit data instead of quota data:
- `X-RateLimit-Limit`: the configured rate limit threshold
- `X-RateLimit-Remaining`: remaining requests in the current rate limit window
- `X-RateLimit-Reset`: time until the rate limit window resets

Backward compatibility is preserved: when the option is unset, the headers continue to reflect quota data as before.
</Accordion>

> <Accordion title='Simplified Management of Session Lifetime'>

Reviewer: ```suggestion
<Accordion title='Simplified management of Session lifetime'>
```

We have added a new, simplified, approach to configuring [Session lifetime](/api-management/access-control/sessions-and-keys/session-lifecycle) within Redis.

Two new fields have been added to the Session object, which can be directly configured when creating Sessions using the [Keys API](/api-reference/keys/create-a-key-1) or [Policy API](/api-reference/policies/create-a-policy):
- [post_expiry_action](/tyk-oss-gateway/configuration#post_expiry_action) - determines what happens to the data in Redis after the `expires` timestamp is reached.
- [post_expiry_grace_period](/tyk-oss-gateway/configuration#post_expiry_grace_period) - defines how long (in seconds) the Session is kept in Redis after expiration (if the `post_expiry_actions` is to retain the Session)

> 

Reviewer: ```suggestion
```

</Accordion>

<Accordion title='Optional toggle to reuse upstream mTLS certificates for OAuth external service calls'>
When configuring mTLS for both upstream authentication and the OAuth external service integration, the same certificate often had to be configured in two places — once for upstream auth, and again in the External Services OAuth configuration. For deployments using the same certificate for both, this created a redundant configuration that needed to be kept in sync.
This release adds an optional configuration toggle that allows the Gateway to reuse the mTLS certificate configured for upstream connections when making OAuth calls to the external Identity Provider. When the toggle is enabled, the upstream certificate is used for both purposes, removing the need to maintain duplicate configurations.
The toggle is intentionally optional: deployments may legitimately use different certificates for the two endpoints.
When the toggle is unset, the OAuth external service continues to use its own dedicated certificate configuration.
</Accordion>

> <Accordion title='Allow Underscores in Vault and Consul Secret Reference Paths'>

Reviewer: ```suggestion
<Accordion title='Allow underscores in HashiCorp Vault and Consul secret reference paths'>
```

Compression is opt-in per asset type via independent configuration flags: [`storage.enable_api_definition_compression`](/tyk-oss-gateway/configuration#storage-enable_api_definition_compression) (introduced in 5.12.0) for API definitions, and the new [`storage.enable_policy_compression`](/tyk-oss-gateway/configuration#storage-enable_policy_compression) for security policies. Both are disabled by default and remain backward compatible — Gateways transparently load both compressed and uncompressed data from Redis.

The decompression size limit is now configurable via the new [`storage.max_decompressed_size`](/tyk-oss-gateway/configuration#storage-max_decompressed_size) option (default 100MB, minimum 1MB) and is applied independently to API definitions and policies, so each asset type has its own ceiling. As with the original feature, compression and decompression occur only during Gateway reloads, with no impact on the request hot path.
</Accordion>

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> <Accordion title='Add OpenTelemetry Metrics Export'>

Reviewer: ```suggestion
<Accordion title='Add OpenTelemetry metrics export'>
```

**Configuration:** Metrics are configured under the new [opentelemetry.metrics](/tyk-oss-gateway/configuration#opentelemetry-metrics) object with independent control from tracing.

**Default Metrics Exported:**
- **Request Metrics:** `http.server.request.duration`, `tyk.gateway.request.duration`, `tyk.upstream.request.duration`, `tyk.api.requests.total`
- **Go Runtime Metrics:** CPU utilisation and processing time, memory usage breakdown (heap, stack, allocations), active goroutine count and lifecycle, garbage collection duration and frequency, thread count
- **Configuration Metrics:** Loaded APIs and policies count, reload operations and duration

The Go runtime metrics enable teams to correlate system-level resource constraints with API latency patterns, detect potential goroutine or memory leaks early, and right-size infrastructure based on actual resource utilisation.

**Production Controls:** To ensure safe metric collection on the request hot path, the system includes built-in cardinality limits (default: 2,000 combinations per metric) and conditional data loading to prevent performance degradation.
</Accordion>

</AccordionGroup>

##### Fixed

<AccordionGroup>

> <Accordion title='Fix Log Level for Client TLS Certificate Requirement Errors'>

Reviewer: ```suggestion
<Accordion title='Fix log level for Client TLS certificate requirement errors'>
```

We have resolved an issue where "Client TLS certificate is required" errors were logged at `warning` level, creating unnecessary noise in production logs.

Previously, these common client-side authentication failures generated excessive warning-level log entries that could trigger false alerts and obscure more critical issues. The Gateway now logs these authentication failures at `info` level, maintaining security visibility while reducing log noise and alert fatigue for operations teams.
</Accordion>

> <Accordion title='Fix Malformed Responses from Go Plugins Returning Error Status Codes'>

Reviewer: ```suggestion
<Accordion title='Fix malformed responses from Go plugins returning error status codes'>
```

We have resolved an issue where a Go plugin returning an error status code would result in malformed response bodies that concatenated the original plugin response with additional Gateway error messages. 

The Gateway now correctly handles plugin-generated error responses without double-writing headers, ensuring response bodies contain only the payload generated by the plugin and eliminating superfluous warnings in logs.
</Accordion>

> <Accordion title='Fix Query Parameter Handling when Looping Requests Internally to Tyk'>

Reviewer: ```suggestion
<Accordion title='Fix query parameter handling when looping requests internally'>
```

> We have resolved inconsistent query parameter handling in URL rewrites when [internally looping](/advanced-configuration/transform-traffic/looping) using the `tyk://` protocol. Previously, custom query parameters specified in the `rewrite_to` URL were silently dropped, while original request parameters were unexpectedly preserved.

Reviewer: ```suggestion
We have resolved inconsistent query parameter handling in URL rewrites when [internally looping](/advanced-configuration/transform-traffic/looping) using the `tyk://` scheme. Previously, custom query parameters specified in the `rewrite_to` URL were silently dropped, while original request parameters were unexpectedly preserved.
```

**What's Fixed:**
- Query parameters explicitly added to `rewrite_to` URLs are now correctly passed to target APIs
- Control parameters (`method`, `loop_limit`, `check_limits`) are properly consumed and removed
- Behavior now matches URL rewrites using `http://` protocol

<Note>
It is important to note that **query parameters provided with the original request are no longer automatically forwarded**. You must update your [URL rewrite configuration](/transform-traffic/url-rewriting) to explicitly include any required parameters in the `rewrite_to` URL.
</Note>
</Accordion>

> <Accordion title='Fix Client mTLS Authentication Between Tyk Gateways'>

Reviewer: ```suggestion
<Accordion title='Fix client mTLS authentication between Tyk Gateways'>
```

We have resolved an issue where a Tyk Gateway acting as a client (using upstream mTLS) would fail to authenticate against another Tyk Gateway acting as the mTLS server, resulting in `HTTP 403 Forbidden: Client TLS certificate is required` errors.

The Gateway now reliably presents the configured upstream client certificate whenever requested by the target server, ensuring seamless mTLS communication between APIs hosted on different Tyk Gateways.
</Accordion>

> <Accordion title='Inherit Maximum (1.3) and Minimum (1.2) TLS Versions from Go'>

Reviewer: ```suggestion
<Accordion title='Inherit maximum (1.3) and minimum (1.2) TLS versions from Go'>
```

We have resolved an issue where the Tyk Gateway default maximum TLS version was incorrectly set to TLS 1.2 instead of TLS 1.3.

Tyk Gateway now follows Go's native TLS defaults (TLS 1.2 minimum, TLS 1.3 maximum), aligning with industry security standards. This maintains full backward compatibility for existing deployments that explicitly configure TLS versions.

To change the maximum TLS version, you must explicitly set [TYK_GW_HTTPSERVEROPTIONS_MAXVERSION](/tyk-oss-gateway/configuration#http_server_options-max_version) for client-to-Gateway connections or [TYK_GW_PROXYSSLMAXVERSION](/tyk-oss-gateway/configuration#proxy_ssl-max_version) for Gateway-to-upstream connections.

To change the minimum TLS version, you must explicitly set [TYK_GW_HTTPSERVEROPTIONS_MINVERSION](/tyk-oss-gateway/configuration#http_server_options-min_version) for client-to-Gateway connections or [TYK_GW_PROXYSSLMINVERSION](/tyk-oss-gateway/configuration#proxy_ssl-min_version) for Gateway-to-upstream connections.

For full details of TLS version configuration see [here](/api-management/implement-tls#controlling-tls-version-&-cipher-suites).
</Accordion>

> <Accordion title='Fix CORS Preflight Blocking by AllowList Middleware'>

Reviewer: ```suggestion
<Accordion title='Fix CORS preflight blocking by AllowList middleware'>
```

We have resolved an issue where CORS preflight OPTIONS requests were incorrectly blocked by the AllowList middleware when `options_passthrough` was disabled.

Previously, when APIs had CORS enabled with Tyk handling OPTIONS requests internally (`options_passthrough: false`), preflight requests would fail AllowList validation because users typically don't explicitly define OPTIONS endpoints in their AllowList configurations, causing "Requested endpoint is forbidden" errors.

The Tyk Gateway now properly recognizes CORS preflight requests and allows them to bypass AllowList middleware checks when Tyk is configured to handle OPTIONS internally, restoring the expected behavior where CORS preflight handling works automatically without requiring explicit OPTIONS endpoint definitions.
</Accordion>

> <Accordion title='Fix Analytics Generation for Tyk OAS API Mock Endpoints'>

Reviewer: ```suggestion
<Accordion title='Fix analytics generation for Tyk OAS API mock endpoints'>
```

We have resolved an issue where Tyk OAS APIs with mock endpoints stopped generating analytics data. This functionality was inadvertently broken while fixing an unrelated internal API proxying issue in Tyk Gateway 5.8.6.

Note that analytics are not generated for mock endpoints in Tyk Classic APIs as has always been the case.
</Accordion>

> <Accordion title='Fix OpenTelemetry Configuration File Settings Being Ignored'>

Reviewer: ```suggestion
<Accordion title='Fix OpenTelemetry configuration file settings being ignored'>
```

We have resolved an issue where OpenTelemetry settings could only be set using environment variables and not the Gateway configuration file (`tyk.conf`).

Now OpenTelemetry can be configured via the [`opentelemetry`](/tyk-oss-gateway/configuration#opentelemetry) section in the Gateway config file (including `enabled`, `exporter`, and `endpoint` fields) or their equivalent environment variables.
</Accordion>

> <Accordion title='Resolved Gateway Entering an Unresponsive State During Startup'>

Reviewer: ```suggestion
<Accordion title='Resolved issue with Gateway entering an unresponsive state during startup'>
```

We have fixed an issue where the Gateway could fail to load APIs and policies if the Control Plane database was temporarily unavailable during startup (either directly or via MDCB). The Gateway will now automatically retry loading configurations with exponential backoff until successful, restoring self-healing capabilities without requiring a manual restart.
</Accordion>

> <Accordion title='Add Strict Validation for Policy ID Characters'>

Reviewer: ```suggestion
<Accordion title='Add strict validation for Policy ID characters'>
```
